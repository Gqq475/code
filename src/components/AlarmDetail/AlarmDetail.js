import React from 'react'
import { assign } from 'lodash'
import { connect } from 'react-redux'
import classes from './AlarmDetail.scss'
import AlarmCommentEditor from 'components/AlarmCommentEditor'
import AlarmNotifyDialog from 'components/AlarmNotifyDialog'
import Utils from '../../Utils.js'

type Props = {
  data: Object,
  id: number,
  token: String,
  userRole: Array
};

export class TextField {
  constructor (key, value) {
    this.key = key
    this.value = value
  }

  getRender (btn) {
    return (
      <div className={classes.textField}>
        <div className={classes.key}>
          {this.key}
        </div>
        <div className={classes.value}>
          {this.value}
          {btn}
        </div>
      </div>
    )
  }
}

export class AlarmDetail extends React.Component {
  props: Props;

  constructor (props) {
    super(props)
    this.findById = this.findById.bind(this)
    this.onEdit = this.onEdit.bind(this)
    this.onSubmitNewComment = this.onSubmitNewComment.bind(this)
    this.toggleNotifyDialog = this.toggleNotifyDialog.bind(this)
    this.state = {
      editing: false,
      newComment: '',
      notifyDialogIsOpen: false
    }
  }

  findById (id) {
    return this.props.data[id]
  }

  onEdit () {
    this.setState(assign({}, this.state, {
      editing: !this.state.editing
    }))
  }

  onSubmitNewComment (comment) {
    this.setState(assign({}, this.state, {
      editing: !this.state.editing,
      newComment: comment
    }))
  }

  toggleNotifyDialog () {
    this.setState(assign({}, this.state, {
      notifyDialogIsOpen: !this.state.notifyDialogIsOpen
    }))
  }

  checkIfUpdateNewComment (selectedData) {
    if (this.state.newComment !== '') {
      selectedData['comments'] = this.state.newComment
      this.state.newComment = ''
    }

    return selectedData
  }

  componentWillUpdate () {
    this.state.editing = false
  }

  render () {
    let selectedData
    let createTime
    let calDuration
    let lastUpdate
    if (this.props.data && this.props.data.length) {
      selectedData = this.findById(this.props.id)
      selectedData = this.checkIfUpdateNewComment(selectedData)

      createTime = Utils.formatDate(selectedData['createTime'])
      calDuration = `${Utils.formatDate(selectedData['calStartTime'])} -
        ${Utils.formatDate(selectedData['calEndTime'])}`
      lastUpdate = Utils.formatDate(selectedData['updateTime'])
    }
    return (
      <div className={`pt-card pt-elevation-2 ${classes['AlarmDetail-container']}`}>
        {
        selectedData && (
          <section>
            <section className={classes.header}>告警详情</section>
            <div className={classes.content}>
              <div className={classes.left}>
                {new TextField(_.ID, selectedData['id']).getRender()}
                {new TextField(_.ALARMTYPE, selectedData['type']).getRender()}
                {new TextField(_.ALARMLEVEL, selectedData['level']).getRender()}
                {new TextField(_.STATUS, selectedData['status']).getRender()}
                {new TextField(_.CREATETIME, createTime).getRender()}
              </div>
              <div className={classes.center}>
                {new TextField(_.PRODUCTID, selectedData['productCode']).getRender()}
                {new TextField(_.LINEID, selectedData['lineCode']).getRender()}
                {new TextField(_.OPERATIONID, selectedData['operationCode']).getRender()}
                {new TextField(_.SUBENTITY, selectedData['subEqCode']).getRender()}
                {new TextField(_.REAL, selectedData['realValue']).getRender()}
                {new TextField('计算时间', calDuration).getRender()}
                {new TextField(_.OOC, selectedData['ooc']).getRender()}
                {new TextField(_.OOS, selectedData['oos']).getRender()}
              </div>
              <div className={classes.right}>
                {new TextField(_.UPDATEBY, selectedData['updateBy']).getRender()}
                {new TextField(_.LASTUPDATE, lastUpdate).getRender()}
                {
                  (Utils.identifyRole(this.props.userRole) !== 'guest-only') ? (
                    <div>
                      {!this.state.editing && new TextField('意见',
                          selectedData['comments']).getRender(
                            <button onClick={this.onEdit}
                              className={`pt-button pt-minimal pt-icon-edit ${classes.editBtn}`} />
                        )}
                      {this.state.editing && (
                        <AlarmCommentEditor
                          id={selectedData['id']}
                          onSubmitNewComment={this.onSubmitNewComment}
                          content={selectedData['comments']} />
                        )}
                      {
                        <button
                          type='button'
                          style={{
                            display: 'block',
                            margin: '0 auto'
                          }}
                          className='pt-button pt-large pt-intent-primary'
                          onClick={this.toggleNotifyDialog}
                          >Notify</button>
                      }
                      <AlarmNotifyDialog
                        token={this.props.token}
                        id={selectedData.id}
                        alarmType={selectedData.type}
                        notifyDialogIsOpen={this.state.notifyDialogIsOpen}
                        toggleNotifyDialog={this.toggleNotifyDialog} />
                    </div>
                  ) : (
                    <div className={classes.textField}>
                      <div className={classes.key}>
                        {'意见'}
                      </div>
                      <div className={classes.value}>
                        {selectedData['comments']}
                      </div>
                    </div>
                  )
                }
              </div>
            </div>
          </section>
        )
      }
      </div>
    )
  }
}

export default connect(
  (state) => ({
    userRole: state.auth.userRole
  })
)(AlarmDetail)
