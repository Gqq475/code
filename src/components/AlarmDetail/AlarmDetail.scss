$border: solid 1px #d7d7d7;

.AlarmDetail-container {
  margin-top: 20px;
  border: $border;
  border-radius: 4px;
  padding: 20px 20px 0;

  .header {
    margin: -20px -20px 0;
    padding: 15px 20px 10px;
    background: #F7F7F7;
    border-bottom: solid 1px #d7d7d7;
  }
}

.content {
  display: flex;
  justify-content: flex-start;

  .left {
    width: 25%;
    padding-top: 20px;
  }

  .center {
    width: 50%;
    border-right: $border;
    border-left: $border;
    padding: 20px 0;
  }

  .right {
    width: 25%;
    padding-top: 20px;
  }
}

.textField {
  padding: 5px 0;
  display: flex;
  justify-content: center;
  align-items: flex-start;

  .key {
    color: #909090;
    width: 40%;
    padding-right: 20px;
    text-align: right;
  }

  .value {
    font-weight: bold;
    color: 333;
    width: 60%;
    word-break: break-all;
    overflow:hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  .editBtn {
    display: block;
  }
}

.notifyBtn {
  display: block;
  margin: 0 auto;
}
