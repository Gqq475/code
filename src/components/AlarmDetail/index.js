import AlarmDetail from './AlarmDetail'
export default AlarmDetail
