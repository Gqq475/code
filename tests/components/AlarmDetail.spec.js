import React from 'react'
import { AlarmDetail } from 'components/AlarmDetail/AlarmDetail'
describe('(Component) AlarmDetail', () => {
  it('render normally if data was not empty', () => {
    const data = [
      'data': {
        'length': 1,
        'createTime': 'May 16, 2017 4:34:36 AM',
        'id': 256,
        'productCode': 'SV509-CF',
        'operationCode': '2100',
        'lineCode': 'PIPR01',
        'subEqCode': 'PIN01',
        'type': 'alarm',
        'condi': '',
        'ooc': 5.0,
        'oos': 8.0,
        'level': 2,
        'realValue': '3片，平均值12.3333',
        'calStartTime': 'May 16, 2017 4:34:36 AM',
        'calEndTime': 'May 16, 2017 4:34:36 AM',
        'summary': '',
        'ocapCode': '',
        'glassCode': 'UY478BN',
        'updateBy': '10118436',
        'comments': '已通知PM吴金原',
        'status': 'new',
        'updateTime': 'May 16, 2017 4:34:36 AM'
      }
    ]

    const wrapper = shallow(
      <AlarmDetail {...data} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('render normally if data was empty', () => {
    const data = [{
      data: {}
    }]

    const wrapper = shallow(
      <AlarmDetail {...data} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  it('test findById function is normal', () => {
    const findById = jest.fn()
    expect(findById()).not.toBeNull()
  })
})
