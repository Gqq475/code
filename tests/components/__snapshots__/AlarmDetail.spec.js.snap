// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) AlarmDetail render normally if data was empty 1`] = `
<div
  className="pt-card pt-elevation-2 AlarmDetail-container"
/>
`;

exports[`(Component) AlarmDetail render normally if data was not empty 1`] = `
<div
  className="pt-card pt-elevation-2 AlarmDetail-container"
/>
`;
